Add initializer lists to the parser

gligneul opened the pull request:
This is a proposal to change the syntax to create records.
This also have a impact on arrays, since `local x = {}` will be an error instead of an array of integers.
See the modifications in the manual for more details.

gligneul committed:
add tests for iniitalizer lists syntax

gligneul committed:
add initalizer lists to the parser

- rename Exp_Table to Ext_InitList
- rename InitList.exps to InitList.fields
- Add type Field

gligneul committed:
add initializer lists to the doc

Also: improve the records documentation.

hishamhm commented on doc/manual.md:
```diff
@@ -55,29 +53,27 @@ return values for functions.
+You can create records with initializer lists, assigning a value to each field
+either by it's name or positional order.
+For instance, you could initalize an instance of the record `Point` with: `{ x =
+3, y = 5 }`, `{ y = 5, x = 3 }` or `{ 3, 5 }`.
```
:-1: to `{ 3, 5 }` — very misleading!

hugomg replied:
I think it is important that it be possible to instantiate records without needing to type all the field names. I don't have a strong preference between a Lua-like `local p = Point.new(3,5)` or a C-like `local p : Point = {3, 5}` though. Just don't make me write `local p : Point = Point.new(3,5)`  :)

mascarenhas replied:
I agree with @hishamhm . If you want positional having `local p = Point.new(3,5)` is good enough for me (and supported by Titan's limited type inference). Having `local p: Point = { 3, 5 }` opens up the question of what things like `local p: Point = { x = 3, 5 }`, `local p: Point = { y = 5, 3 }`, or `local p: Point = { 5, x = 3 }` mean...

gligneul replied:
Removed the `{ 3, 5 }` syntax for records. Now we have different ast nodes for arrays constructors and records initalizer lists, so things like `{ y = 5, 3 }` are syntax errors.

hugomg commented on doc/manual.md:
```diff
@@ -55,29 +53,27 @@ return values for functions.
+Like arrays, Titan will try to infer the type of the initializer list, but it is
+necessary to specify the type of the record when declaring a new variable
+(`local p: Point = { 3, 5 }`).
+
```
I don't know if this is out of scope for this PR but currently our type inference rules are undocumented and untested. I think we should rectify that if we are going to strongly rely on type inference for record instantiation.

gligneul replied:
I don't know how to express that without being repetitive (there is already an explanation on the array section). Maybe we should have a section just for the inference rules, but that is indeed beyond the scope of this PR.

gligneul committed:
bring the syntax `Type.new` back to the manual.

Also undo the changes to the Array section.

gligneul committed:
create separate nodes for array constructors and initalizer lists

Now we have the ast node Exp_ArrCons for `{ exp1, exp2, ... }` and
Exp_InitList for `{ field1 = exp1, field2 = exp2, ... }`.

gligneul committed:
add test case that doesn't allow mixed array constructors and initalizer lists

hugomg reviewed (changes requested):
Turning `{x=11,  22}` into a syntax error will stop Titan from creating plain Lua tables with mixed array and hash parts. We currently don't implement this feature but I think it would be very desirable in the future for Lua interoperability.

I think the type checker would be a better place to catch these array and record initialization errors.

gligneul committed:
Revert "add test case that doesn't allow mixed array constructors and initalizer lists"

This reverts commit c4dc9a7215e651e6be96d4d1e6ebd6cbdd25c117.

gligneul committed:
Revert "create separate nodes for array constructors and initalizer lists"

This reverts commit 0c8e18789385b125b1cfa6f60cc98644c60020c1.

mascarenhas commented:
> Turning {x=11, 22} into a syntax error will stop Titan from creating plain Lua tables with mixed array and hash parts. We currently don't implement this feature but I think it would be very desirable in the future for Lua interoperability.

Just because it is a syntax error now does not mean it must be always a syntax error! If it is easy enough to forbid in the parser I do not see a need to make the typechecker more complex right now for the sake of a future feature.

gligneul commented:
> Just because it is a syntax error now does not mean it must be always a syntax error! If it is easy enough to forbid in the parser I do not see a need to make the typechecker more complex right now for the sake of a future feature.

Even tough I agree with you, the changes in the typechecker might be simpler than the changes I've made to the parser. Don't merge it yet, let me check first please.

gligneul committed:
add test case for `Record.new` constructors

gligneul committed:
typecheck the assignment of initlist to array
